File: simulation/test4.java
import java.util.Scanner;


public class test4{

	public static void main(String[] args){

		Scanner sc= new Scanner(System.in);
		int n = sc.nextInt();
		int m = sc.nextInt();
		int posX = sc.nextInt();
		int posY = sc.nextInt();
		int dir = sc.nextInt();	
		int count=1;
		int[][] map = new int[n][m];
		boolean[][] check = new boolean[n][m];
		for(int i=0; i<n; i++){
			for(int j=0 ; j<m; j++){
				check[i][j]=false;
				map[i][j]=sc.nextInt();
			}
		}
		check[posX][posY]=true;
		int[] dirX={-1,0, 1,0};
		int[] dirY={ 0,-1,0,1};
		int turn=0;
		while(true){
			if(dir==4){
				dir=0;
			}
			
			if(!check[posX+dirX[dir]][posY+dirY[dir]] && map[posX+dirX[dir]][posY+dirY[dir]]==0){
				turn=0;
				count++;
				posX+=dirX[dir];
				posY+=dirY[dir];
				check[posX][posY]=true;
				continue;
			}else{	
				dir++;
				turn++;
			}

			if(turn==4){
				if(map[posX-dirX[dir]][posY-dirY[dir]]==1){	
					break;
				}else{
					posX-=dirX[dir];
					posY-=dirY[dir];
					turn=0;
				}
				
			}
		}
		System.out.println(count);

	}


}

File: simulation/test4.py
n,m = map(int, input().split())

check = [ [False] * m for _ in range(n)]
x, y, dir = map(int, input().split())
check[x][y]=True

arr=[]
for i in range(n):
	arr.append(list(map(int, input().split())))

dx=[0,-1,0,1]
dy=[-1,0,1,0]
turn=0
count=1
while True:
	if dir == 4 :
		dir=0
	
	if check[x+dx[dir]][y+dy[dir]]==False and arr[x+dx[dir]][y+dy[dir]] ==0 :
		x= x+dx[dir]
		y= y+dy[dir]
		count+=1
		turn=0
		check[x][y]=True
		continue
	else :
		turn+=1
		dir+=1
	

	if turn==4:
		if arr[x-dx[dir]][y-dy[dir]] == 1:
			break
		else :
			x = x-dx[dir]
			y = y-dy[dir]
			turn = 0



print(count)

